[staging-next] linuxPackages.bcc: fix build

r-burns opened the pull request:
Use Debian downstream patch to fix compilation against linuxHeaders < 5.14.

Our linux headers exposed by glibc are still at 5.12. We should be able
to remove this patch once our linuxHeaders are bumped to 5.14+.

###### Motivation for this change

Fixes linuxPackages*.bcc
For https://github.com/NixOS/nixpkgs/pull/138268

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

martinetd commented:
I'd be tempted to say we just don't need btf kind float on nixos kernels (I had the same problem on another distro so have a patch to rip it off somewhere...), but they don't have btf enabled anyway so I'm assuming you're compiling your own kernels and then anything goes wrt. target version and you might need it...

Ultimately upstream libbpfland prefers embedding while distros prefer using distro's version of the lib so I don't really care either way, but I also don't see the harm in just upgrading glibc's kernel headers to 5.14 -- kernel headers are part of the userspace api which is guaranted to be stable, so there's no harm in being ahead of the running kernel version.

So /shrug, leaving it up to whoever can take this kind of decisions.

r-burns commented:
Yeah, that all sounds reasonable to me, I'd welcome a more principled fix. If we do decide bumping glibc's linux headers is the best solution, it would rebuild the universe and require a lot of testing, so I think we'd still want to merge this in the interim.

martinetd commented:
It makes sense to have this in the meantime, I'm open to anything. I think this kind of problem will unfortunately keep poping up regularly...

I was working on a pahole/bcc/bpftrace update recently and forgot about it, so happy if others help with this stuff :)

Related: here's what debian did for this particular problem:
https://sources.debian.org/src/bpfcc/0.21.0+ds-1/debian/patches/0004-compat-defs.patch/

that might be another solution?
afaiu libbpf 0.4 itself compiles fine, and it's a bcc problem?

r-burns commented:
Ooh, nice find, tried it out and it seems to work. Maybe I should update this PR to use that instead?

r-burns committed:
linuxPackages.bcc: fix build

Use Debian downstream patch to fix compilation against linuxHeaders < 5.14.

Our linux headers exposed by glibc are still at 5.12. We should be able
to remove this patch once our linuxHeaders are bumped to 5.14+.

r-burns commented:
Result of `nixpkgs-review pr 138465` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages marked as broken and skipped:</summary>
  <ul>
    <li>linuxKernel.packages.linux_latest_libre.bcc</li>
    <li>linuxKernel.packages.linux_latest_libre.bpftrace</li>
    <li>linuxKernel.packages.linux_latest_libre.oci-seccomp-bpf-hook</li>
    <li>linuxKernel.packages.linux_libre.bcc</li>
    <li>linuxKernel.packages.linux_libre.bpftrace</li>
    <li>linuxKernel.packages.linux_libre.oci-seccomp-bpf-hook</li>
  </ul>
</details>
<details>
  <summary>45 packages built:</summary>
  <ul>
    <li>linuxKernel.packages.linux_4_14.bcc</li>
    <li>linuxKernel.packages.linux_4_14.bpftrace</li>
    <li>linuxKernel.packages.linux_4_14_hardened.bcc</li>
    <li>linuxKernel.packages.linux_4_14_hardened.bpftrace</li>
    <li>linuxKernel.packages.linux_4_19.bcc</li>
    <li>linuxKernel.packages.linux_4_19.bpftrace</li>
    <li>linuxKernel.packages.linux_4_19_hardened.bcc</li>
    <li>linuxKernel.packages.linux_4_19_hardened.bpftrace</li>
    <li>linuxKernel.packages.linux_4_4.bcc</li>
    <li>linuxKernel.packages.linux_4_4.bpftrace</li>
    <li>linuxKernel.packages.linux_4_9.bcc</li>
    <li>linuxKernel.packages.linux_4_9.bpftrace</li>
    <li>linuxKernel.packages.linux_5_10.bcc</li>
    <li>linuxKernel.packages.linux_5_10.bpftrace</li>
    <li>linuxKernel.packages.linux_5_10.oci-seccomp-bpf-hook</li>
    <li>linuxKernel.packages.linux_hardened.bcc (linuxKernel.packages.linux_5_10_hardened.bcc)</li>
    <li>linuxKernel.packages.linux_hardened.bpftrace (linuxKernel.packages.linux_5_10_hardened.bpftrace)</li>
    <li>linuxKernel.packages.linux_hardened.oci-seccomp-bpf-hook (linuxKernel.packages.linux_5_10_hardened.oci-seccomp-bpf-hook)</li>
    <li>linuxKernel.packages.linux_5_13.bcc</li>
    <li>linuxKernel.packages.linux_5_13.bpftrace</li>
    <li>linuxKernel.packages.linux_5_13.oci-seccomp-bpf-hook</li>
    <li>linuxKernel.packages.linux_5_13_hardened.bcc</li>
    <li>linuxKernel.packages.linux_5_13_hardened.bpftrace</li>
    <li>linuxKernel.packages.linux_5_13_hardened.oci-seccomp-bpf-hook</li>
    <li>linuxKernel.packages.linux_5_14.bcc</li>
    <li>linuxKernel.packages.linux_5_14.bpftrace</li>
    <li>linuxKernel.packages.linux_5_14.oci-seccomp-bpf-hook</li>
    <li>linuxKernel.packages.linux_5_4.bcc</li>
    <li>linuxKernel.packages.linux_5_4.bpftrace</li>
    <li>linuxKernel.packages.linux_5_4.oci-seccomp-bpf-hook</li>
    <li>linuxKernel.packages.linux_5_4_hardened.bcc</li>
    <li>linuxKernel.packages.linux_5_4_hardened.bpftrace</li>
    <li>linuxKernel.packages.linux_5_4_hardened.oci-seccomp-bpf-hook</li>
    <li>linuxKernel.packages.linux_lqx.bcc</li>
    <li>linuxKernel.packages.linux_lqx.bpftrace</li>
    <li>linuxKernel.packages.linux_lqx.oci-seccomp-bpf-hook</li>
    <li>linuxKernel.packages.linux_testing_bcachefs.bcc</li>
    <li>linuxKernel.packages.linux_testing_bcachefs.bpftrace</li>
    <li>linuxKernel.packages.linux_testing_bcachefs.oci-seccomp-bpf-hook</li>
    <li>linuxKernel.packages.linux_xanmod.bcc</li>
    <li>linuxKernel.packages.linux_xanmod.bpftrace</li>
    <li>linuxKernel.packages.linux_xanmod.oci-seccomp-bpf-hook</li>
    <li>linuxKernel.packages.linux_zen.bcc</li>
    <li>linuxKernel.packages.linux_zen.bpftrace</li>
    <li>linuxKernel.packages.linux_zen.oci-seccomp-bpf-hook</li>
  </ul>
</details>

martinetd commented on pkgs/os-specific/linux/bcc/default.nix:
```diff
@@ -29,6 +29,12 @@ python.pkgs.buildPythonApplication rec {
+    # Add definition for BTF_KIND_FLOAT, added in Linux 5.14
+    # Can be removed once linuxHeaders (used here via glibc) are bumped to 5.14+.
+    (fetchpatch {
+      url = "https://salsa.debian.org/debian/bpfcc/-/raw/master/debian/patches/0004-compat-defs.patch";
```
point to the commit id of current master instead of master, in case debian removes the patch or its content changes? (build will break with checksum, but if we can make build not break instead...)

r-burns replied:
Oops good catch. Obviously should be a permalink.

martinetd reviewed (approved):
Thanks!

vcunat commented:
Headers update is in PR #136358; that surely won't make it into the current staging-next iteration anymore.

Mic92 commented:
Let's revert this in the mentioned PR.
